system_metrics_collector/src/topic_statistics_collector/subscriber_topic_statistics.hpp

```diff
@@ -0,0 +1,373 @@
+    RCLCPP_DEBUG(this->get_logger(), "on_deactivate");
+    for (const auto & collector : statistics_collectors_) {
+      if (!collector->Stop()) {
+        return CallbackReturn::ERROR;
+      }
+    }
+
+    publish_timer_->cancel();
+    publisher_->on_deactivate();
+    return CallbackReturn::SUCCESS;
+  }
+
+  /**
+   * Stops the node and performs cleanup.
+   *
+   * @param input state unused
+   * @return CallbackReturn success
+   */
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_shutdown(
+    const rclcpp_lifecycle::State & state)
+  {
+    RCLCPP_DEBUG(this->get_logger(), "on_shutdown");
+    StopPublisher();
+    publisher_.reset();
+
+    return CallbackReturn::SUCCESS;
+  }
+
+  /**
+   * Stops the node and attempts to perform cleanup.
+   *
+   * @param input state unused
+   * @return CallbackReturn success
+   */
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_error(
+    const rclcpp_lifecycle::State & previous_state)
+  {
+    RCLCPP_DEBUG(this->get_logger(), "on_error");
+    StopPublisher();
+    if (publisher_ != nullptr) {
+      publisher_.reset();
+    }
+
+    return CallbackReturn::SUCCESS;
+  }
+
+protected:
+  /**
+   * LifecyclePublisher publisher that periodically publishes statistic messages
+   *
+   */
+  rclcpp_lifecycle::LifecyclePublisher<metrics_statistics_msgs::msg::MetricsMessage>::SharedPtr
+    publisher_;
+
+  /**
+   * ROS2 topic statistics collectors
+   */
+  std::vector<std::unique_ptr<TopicStatisticsCollector<T>>> statistics_collectors_;
+
+private:
+  /**
+   * Creates ROS2 timers and a publisher for periodically triggering measurements
+   * and publishing MetricsMessages
+   */
+  void StartPublisher()
+  {
+    if (publisher_ == nullptr) {
+      publisher_ = create_publisher<metrics_statistics_msgs::msg::MetricsMessage>(
+        publish_topic_name_,
+        10);
+    }
+
+    publisher_->on_activate();
+
+    RCLCPP_DEBUG(this->get_logger(), "SetupStart: creating publish_timer_");
+    publish_timer_ = this->create_wall_timer(
+      publish_period_, [this]() {
+        this->PublishStatisticMessage();
+        this->ClearCollectorMeasurements();
+      });
+    window_start_ = now();
+  }
+
+  /**
+   * Stops the ROS2 timers that were created by StartPublisher()
+   */
+  void StopPublisher()
+  {
+    rcpputils::check_true(publish_timer_ != nullptr);
+    rcpputils::check_true(publisher_ != nullptr);
+
+    ClearCollectorMeasurements();
+    publisher_->on_deactivate();
+
+    publish_timer_->cancel();
+    publish_timer_.reset();
+  }
+
+  /**
+  * Clear the emasurements made by all statistics collectors
+  */
+  void ClearCollectorMeasurements()
+  {
+    for (const auto & collector : statistics_collectors_) {
+      collector->ClearCurrentMeasurements();
+    }
+  }
+
+  /**
+   * Publishes the statistics derived from the collected measurements (this is to be called via a
+   * ROS2 timer per the publish_period)
+   */
```

Comment:
Nit: Add `@throws rcpputils::IllegalStateException if publisher is null or publisher_ is not activated`